File: backend/db/models/prediction.py
import uuid
from datetime import datetime

from sqlalchemy import Column, DateTime, ForeignKey, Integer, String, UniqueConstraint
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship

from db.database import Base


class Prediction(Base):
    __tablename__ = "predictions"
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id"))
    league_id = Column(UUID(as_uuid=True), ForeignKey("leagues.id"))
    created_at = Column(DateTime, default=datetime.utcnow)
    shared_slug = Column(String, unique=True)

    user = relationship("User", back_populates="predictions")
    league = relationship("League", back_populates="predictions")
    entries = relationship("PredictionEntry", back_populates="prediction")

    __table_args__ = (
        UniqueConstraint('user_id', 'league_id', name='_user_league_uc'),
    )


class PredictionEntry(Base):
    __tablename__ = "prediction_entries"
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    prediction_id = Column(UUID(as_uuid=True), ForeignKey("predictions.id"))
    team_id = Column(UUID(as_uuid=True), ForeignKey("teams.id"))
    predicted_position = Column(Integer, nullable=False)

    prediction = relationship("Prediction", back_populates="entries")
    team = relationship("Team", back_populates="prediction_entries")

File: tipssvenskan/lib/tipssvenskan/predictions/prediction.ex
defmodule Tipssvenskan.Predictions.Prediction do
  use Ecto.Schema
  import Ecto.Changeset

  schema "predictions" do
    field :shared_slug, :string

    belongs_to :user, Tipssvenskan.Accounts.User
    belongs_to :league, Tipssvenskan.Leagues.League
    has_many :entries, Tipssvenskan.Predictions.PredictionEntry

    timestamps(type: :utc_datetime)
  end

  def changeset(prediction, attrs) do
    prediction
    |> cast(attrs, [:shared_slug, :user_id, :league_id])
    |> validate_required([:shared_slug, :user_id, :league_id])
    |> unique_constraint(:shared_slug)
    |> unique_constraint([:user_id, :league_id])
  end
end
